ios/brave-ios/Sources/PlaylistUI/PlayerModel.swift

```diff
@@ -0,0 +1,494 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import AVKit
+import Combine
+import Foundation
+import MediaPlayer
+import SwiftUI
+import os
+
+@available(iOS 16.0, *)
+final class PlayerModel: ObservableObject {
+
+  init() {
+    // FIXME: Consider moving this to a setUp method and call explicitly from UI
+    player.seek(to: .zero)
+    player.actionAtItemEnd = .none
+
+    playerLayer.videoGravity = .resizeAspect
+    playerLayer.needsDisplayOnBoundsChange = true
+    playerLayer.player = player
+
+    setupPlayerKeyPathObservation()
+    setupRemoteCommandCenterHandlers()
+    updateSystemPlayer()
+
+    DispatchQueue.global().async {
+      // should mode be `.moviePlayback`?
+      try? AVAudioSession.sharedInstance().setCategory(.playback, mode: .default)
+    }
+  }
+
+  deinit {
+    if isPlaying {
+      log.warning("PlayerModel deallocated without first stopping the underlying media.")
+      stop()
+    }
+  }
+
+  private let log = Logger(subsystem: Bundle.main.bundleIdentifier!, category: "PlayerModel")
+
+  // MARK: - Player Status
+
+  var isPlaying: Bool {
+    get {
+      player.timeControlStatus == .playing
+    }
+    set {
+      if newValue {
+        play()
+      } else {
+        pause()
+      }
+    }
+  }
+
+  var currentTime: TimeInterval {
+    guard let item = player.currentItem else { return 0 }
+    return max(0, min(duration, item.currentTime().seconds))
+  }
+
+  var duration: TimeInterval {
+    guard let item = player.currentItem else { return 0 }
+    let seconds = CMTimeConvertScale(
+      item.asset.duration,
+      timescale: item.currentTime().timescale,
+      method: .roundHalfAwayFromZero
+    ).seconds
+    return seconds.isNaN ? 0.0 : seconds
+  }
+
+  var currentTimeStream: AsyncStream<TimeInterval> {
+    return .init { [weak self] continuation in
+      guard let self else { return }
+      let observer = player.addCancellablePeriodicTimeObserver(forInterval: 500) { [weak self] _ in
+        guard let self else { return }
+        continuation.yield(currentTime)
+      }
+      // Should be tied to the View, but adding ane extra killswitch
+      self.cancellables.insert(observer)
+      continuation.onTermination = { _ in
+        observer.cancel()
+      }
+    }
+  }
+
+  var didPlayToEndStream: AsyncStream<Void> {
+    return .init { [weak self] continuation in
+      guard let self else { return }
+      let observer = NotificationCenter.default.addObserver(
+        forName: AVPlayerItem.didPlayToEndTimeNotification,
+        object: nil,
+        queue: .main
+      ) { _ in
+        continuation.yield()
+      }
+      let cancellable = AnyCancellable {
+        _ = observer
+      }
+      // Should be tied to the View, but adding ane extra killswitch
+      self.cancellables.insert(cancellable)
+      continuation.onTermination = { _ in
+        cancellable.cancel()
+      }
+    }
+  }
+
+  // MARK: - Player Controls
+
+  func play() {
+    if isPlaying {
```

Comment:
Where are we setting `player.timeControlStatus`? I expected play pause and stop will take care of it.

Reply:
Thats all handled under the hood by `AVPlayer` when you play/pause/etc, we simply observe properties/notifications